### DISCUSSION TITLE
PowerShell Module Packaging Questions

### DISCUSSION THREAD

**Discussion by rhubarb-geek-nz** [PowerShell/PowerShell#19328] [Q&A]
I have a couple of questions regarding packaging of  PowerShell modules.

1. With a C# module what dependencies should be included in the module package?

For example using "dotnet publish --configuration Release" also includes the following;

```
bin/Release/netstandard2.0/publish/Microsoft.Bcl.AsyncInterfaces.dll
bin/Release/netstandard2.0/publish/Microsoft.Bcl.HashCode.dll
bin/Release/netstandard2.0/publish/Microsoft.Extensions.Logging.Abstractions.dll
bin/Release/netstandard2.0/publish/System.Buffers.dll
bin/Release/netstandard2.0/publish/System.Collections.Immutable.dll
bin/Release/netstandard2.0/publish/System.Diagnostics.DiagnosticSource.dll
bin/Release/netstandard2.0/publish/System.Memory.dll
bin/Release/netstandard2.0/publish/System.Numerics.Vectors.dll
bin/Release/netstandard2.0/publish/System.Runtime.CompilerServices.Unsafe.dll
bin/Release/netstandard2.0/publish/System.Text.Encodings.Web.dll
bin/Release/netstandard2.0/publish/System.Text.Json.dll
bin/Release/netstandard2.0/publish/System.Threading.Channels.dll
bin/Release/netstandard2.0/publish/System.Threading.Tasks.Extensions.dll
```

All of these except for Microsoft.Extensions.Logging.Abstractions.dll already exist in PowerShell itself.

Should all these modules be included in the Module package?

2. I want to package and distribute these modules as native packages, eg MSI, OSX pkg, RPM and deb files.

What are the recommended target folders to install the module?

On Linux it is not recommended that packages write into /usr/local.

My analysis leads me to..

| RID | Installation Directory |
| --- | ---------------------- |
| win | c:\Program Files\WindowsPowerShell\Modules |
| win | c:\Program Files\PowerShell\Modules |
| osx | /usr/local/share/powershell/Modules |
| linux | /opt/microsoft/powershell/7/Modules |

Is there any formal guidance?

3. Do PowerShell modules need the deps.json and pdb files?

As seen by the native module loading the deps.json file seems to be ignored for PowerShell modules.

**Comment by jborean93** [comment]
> Should all these modules be included in the Module package?

If pwsh ships those modules then you shouldn't include them. Add `PrivateAssets="all"` to the `PackageReference` in your csproj to ensure they aren't included in the publish folder.

> What are the recommended target folders to install the module?

There is no official recommendation AFAIK because the official recommendation is to ship it as a nuget package and use PowerShellGet to install them.

> Do PowerShell modules need the deps.json and pdb files?

Strictly no, but it doesn't hurt to include them, the pdb file could be useful if the end user needs to debug something in your code but that would most likely be a rare scenario.

**Reply by rhubarb-geek-nz** [reply]
> > Should all these modules be included in the Module package?
> 
> If pwsh ships those modules then you shouldn't include them. Add `PrivateAssets="all"` to the `PackageReference` in your csproj to ensure they aren't included in the publish folder.

https://github.com/rhubarb-geek-nz/NpgsqlConnection

https://github.com/rhubarb-geek-nz/NpgsqlConnection/blob/main/NpgsqlConnection.csproj

I already have that as an element on PowerShellStandard.Library, if I add that to Npgsql then I get no dependencies at all in the publish folder.

At the moment I manually select to add just Microsoft.Extensions.Logging.Abstractions.dll otherwise it fails to load. What it sounds like is I should accept to add those dlls because publish thinks I need them, and I can't rely on manually matching with what any particular PowerShell installation has.

**Reply by jborean93** [reply]
> I already have that as an element on PowerShellStandard.Library, if I add that to Npgsql then I get no dependencies at all in the publish folder.

You would have to add it for each of the deps explicitly AFAIK to have it avoid shipping them. It's not the end of the world if they are included but it gets messy due to assembly version conflicts (next point below).

> What it sounds like is I should accept to add those dlls because publish thinks I need them, and I can't rely on manually matching with what any particular PowerShell installation has.

This starts getting into the territory where you need to use an Assembly Load Context to load your dependencies to avoid conflicts with what PowerShell ships with and what other modules might provide. When you start adding dependencies that could be loaded by something else, PowerShell or other modules, then you might start to encounter loading errors which are very tricky to pin down and figure out.

There's a decent blog about ALC with PowerShell at https://learn.microsoft.com/en-us/powershell/scripting/dev-cross-plat/resolving-dependency-conflicts?view=powershell-7.3 which talks about this problem further. I will have to admit I'm just starting to use this for the first time in one of my modules and it hasn't been simple but it is doable.

**Reply by rhubarb-geek-nz** [reply]
> There is no official recommendation AFAIK because the official recommendation is to ship it as a nuget package and use PowerShellGet to install them.

How does this work to install packages for all users? The last time I tried a similar thing it was to use dotnet to install PowerShell as a global tool, instead it just installed PowerShell for the user root under the /root directory but nobody else could see it or use it.

Does this mechanism support installing a single module once for all users on a system, or is it like the MSIX system which effectively treats installing applications as a per-user feature rather than like MSI which install per machine.

**Reply by jborean93** [reply]
PowerShellGet has 2 scopes:

* CurrentUser (default on 7+)
* AllUsers (default on 5.1)

On non-Windows it uses the following to get those paths

```powershell
# CurrentUser
[System.Management.Automation.Platform]::SelectProductNameForDirectory('USER_MODULES')

# AllUsers
[System.Management.Automation.Platform]::SelectProductNameForDirectory('SHARED_MODULES')
```

The `AllUsers` example on Linux is `/usr/local/share/powershell/Modules`. I would treat `/opt/microsoft/powershell/x/Modules` like the `C:\Windows\System32\WindowsPowerShell\v1.0\Modules` folder and keep it for PowerShell itself to manage. You can still use it if you want to avoid touching `/usr/local/` but it goes against the grain of how PowerShellGet operates.

You can extend the paths used by PowerShell with the `PSModulePath` env var. by default it includes `$PSHome/Modules`, the `AllUsers`, and `CurrentUser` paths but as it's an env var it can be whatever you want.

> Does this mechanism support installing a single module once for all users on a system,

As long as the user has read access to the directory and the dir is in `PSModulePath`, they will be able to load the module automatically.

**Comment by rhubarb-geek-nz** [comment]
@jborean93 great, thank you. I will look at doing the NuGet properly for packaging my modules
